DFC.ServiceTaxonomy.PageLocation/Handlers/PageLocationTaxonomyTermHandler.cs

```diff
@@ -0,0 +1,58 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using DFC.ServiceTaxonomy.PageLocation.Models;
+using DFC.ServiceTaxonomy.Taxonomies.Handlers;
+using DFC.ServiceTaxonomy.Taxonomies.Helper;
+using Newtonsoft.Json.Linq;
+using OrchardCore.ContentManagement;
+using OrchardCore.ContentManagement.Records;
+using YesSql;
+
+namespace DFC.ServiceTaxonomy.PageLocation.Handlers
+{
+    public class PageLocationTaxonomyTermHandler : ITaxonomyTermHandler
+    {
+        private readonly ISession _session;
+        private readonly ITaxonomyHelper _taxonomyHelper;
+        private readonly IContentManager _contentManager;
+
+        public PageLocationTaxonomyTermHandler(ISession session, ITaxonomyHelper taxonomyHelper, IContentManager contentManager)
+        {
+            _session = session;
+            _taxonomyHelper = taxonomyHelper;
+            _contentManager = contentManager;
+        }
+
+        public async Task UpdatedAsync(ContentItem term, ContentItem taxonomy)
+        {
+            if (term.Content.PageLocation == null)
+                return;
+
+            IEnumerable<ContentItem> allPages = await _session.Query<ContentItem, ContentItemIndex>(x => x.ContentType == "Page" && x.Latest).ListAsync();
+            IEnumerable<ContentItem> associatedPages = allPages.Where(x => x.Content.Page.PageLocations.TermContentItemIds[0] == term.ContentItemId);
+
+            foreach (var page in associatedPages)
+            {
+                //rebuild the Full URL to ensure it matches the current state of the term
+                var pageUrlName = page.As<PageLocationPart>().UrlName;
+                var termUrl = _taxonomyHelper.BuildTermUrl(JObject.FromObject(term), JObject.FromObject(taxonomy));
+
+                var fullUrl = string.IsNullOrWhiteSpace(termUrl)
+                    ? $"/{pageUrlName}"
+                    : $"/{termUrl}/{pageUrlName}";
+
+                page.Alter<PageLocationPart>(part => part.FullUrl = fullUrl);
+
+                if (page.Published)
+                {
+                    page.Published = false;
+                    await _contentManager.PublishAsync(page);
+                    return;
+                }
+
+                await _contentManager.SaveDraftAsync(page);
```

Comment:
I think we need to update draft and published pages separately if necessary (including updating both graphs and publishing event grid events), or at least give this some more thought, as I don't think its as simple as re-saving draft pages.

E.g. if an item has a published version and a draft version, this will update the draft page, but the published version of the page will have an url with the old page location, and page locations are published only, so there'd be a mismatch between the published page url and the published locations.

Reply:
I believe the logic in there now handles things appropriately, and the draft/pub graphs seem to keep up to date with the right state

Reply:
Can you please talk me though it, as I don't understand how it handles published and draft items correctly?

It fetches only the latest pages, so if a page has both a published and a draft version, only the the draft page is returned from the query fetching the latest pages, so surely the published version of the page doesn't get updated?